mkosi/distributions/gentoo.py

```diff
@@ -211,125 +118,142 @@ def fetch_fix_stage3(self) -> None:
-        if not stage3_tar_path.is_file():
-            log_step(f"Fetching {stage3_url_path}")
-            stage3_tar_path.parent.mkdir(parents=True, exist_ok=True)
-            urllib.request.urlretrieve(stage3_url_path, stage3_tar_path)
-
-        stage3_tmp_extract.mkdir(parents=True, exist_ok=True)
-
-        with flock(stage3_tmp_extract):
-            if not stage3_tmp_extract.joinpath(".cache_isclean").exists():
-                log_step(f"Extracting {stage3_tar.name} to {stage3_tmp_extract}")
 
+        cls.stage3_cache = state.cache_dir.joinpath("stage3")
+        if not stage3_tar_path.is_file():
+            if cls.stage3_cache.exists():
+                log_step("New stage3 is available, removing the old one")
+                unlink_try_hard(cls.stage3_cache)
+            with complete_step(f"Fetching {stage3_url_path}"):
+                stage3_tar_path.parent.mkdir(parents=True, exist_ok=True)
+                urllib.request.urlretrieve(stage3_url_path, stage3_tar_path)
+        cls.stage3_cache.mkdir(parents=True, exist_ok=True)
+        if not cls.stage3_cache.joinpath(".cache_isclean").exists():
+            with complete_step(f"Extracting {stage3_tar.name} to {cls.stage3_cache}"):
                 run([
                     "tar",
                     "--numeric-owner",
-                    "-C", stage3_tmp_extract,
+                    "-C", cls.stage3_cache,
                     "--extract",
                     "--file", stage3_tar_path,
                     "--exclude", "./dev",
+                    "--exclude", "./proc",
                 ])
 
-                unlink_try_hard(stage3_tmp_extract.joinpath("dev"))
-                unlink_try_hard(stage3_tmp_extract.joinpath("proc"))
-                unlink_try_hard(stage3_tmp_extract.joinpath("sys"))
-
-                stage3_tmp_extract.joinpath(".cache_isclean").touch()
-
-        log_step(f"Copying {stage3_tmp_extract} to {self.root}")
-        copy_path(stage3_tmp_extract, self.root)
-
-    def set_useflags(self) -> None:
-        package_use = self.portage_cfg_dir / "package.use"
-        package_use.mkdir(exist_ok=True)
-
-        package_use.joinpath("systemd").write_text(
-            # repart for usronly
-            dedent(
-                """\
-                # MKOSI: used during the image creation
-                # "/usr/lib/systemd/boot/efi": No such file or directory
-                sys-apps/systemd gnuefi
-                sys-apps/systemd -cgroup-hybrid
-                sys-apps/systemd elfutils # for coredump
-
-                sys-apps/systemd homed cryptsetup -pkcs11
-                # See: https://bugs.gentoo.org/832167
-                sys-auth/pambase homed
-
-                # MKOSI: usronly
-                sys-apps/systemd repart
-                # MKOSI: make sure we're init (no openrc)
-                sys-apps/systemd sysv-utils
-                """
+            for d in ("binpkgs", "distfiles", "repos"):
```

Comment:
Everything from this point onwards doesn't need to be cached as it will be more or less instant to execute.